File: 0098. 验证二叉搜索树/cpp/solution.cpp

#include <climits>
#include <iostream>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right)
        : val(x), left(left), right(right) {}
};

class Solution {
   public:
    bool dfs(TreeNode *root, long long low, long long high) {
        if (root == nullptr) return true;
        if (root->val <= low || root->val >= high) return false;
        return dfs(root->left, low, root->val) &&
               dfs(root->right, root->val, high);
    }
    bool isValidBST(TreeNode *root) { return dfs(root, LONG_MIN, LONG_MAX); }
};

int main() { return 0; }

File: 0098. 验证二叉搜索树/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function (root) {
  const traversal = (node, min, max) => {
    if (node === null) return true
    if (node.val <= min || node.val >= max) return false
    const left = traversal(node.left, min, node.val)
    const right = traversal(node.right, node.val, max)
    return left && right
  }
  return traversal(root, -Infinity, Infinity)
}
